### ISSUE TITLE
limiting memory usage

### ISSUE THREAD

**Issue by tomek1024** [bensheldon/good_job#1170]
Before we moved to good_job we used to use resque. It works in a different way than good_job, it spawns a child process for every task from the queue. It's slow but convenient when some tasks require a lot of RAM. After tasks are finished, the memory is always released to the operating system.

We use good job as an external process with a few threads. After migrating, we have fixed and speeded up some tasks, but still, it sometimes happens, that they grow and use too much memory. We use docker which monitors if the process is running. If not, it will be started again. I would like to be able to set up a max of consuming memory. If too much memory has been used, the worker could gracefully shut down and wait to be started again by docker or any other monitoring tool.

What do you think about it?

**Comment by tomek1024** [comment]
Thank you very much for your very fast answer! I think you are right suggesting, that this should be solved outside of a running good job process. We have a similar problem with puma, and we come to the following solution:

1. we check with `docker stats` the memory usage and ID of the puma container
2. If the memory usage is too high we just call `docker exec -it <ID> kill -SIGUSR1 1` (this first docker process always has PID 1)

This way we restart puma gracefully, without losing any request. We could process similar with good job. but instead of just killing the process I would prefer to send him a SIGUSER1 or SIGUSER2, so that good job process stops accepting new jobs, finishes all the running tasks and then shuts down. Some monitoring system would make start it again. Is such a solution possible today?

**Comment by tomek1024** [comment]
I've made a test with an artificially slow job - after sending a TERM signal the process waits until it finishes :)

**Comment by bensheldon** [comment]
Yep, by default GoodJob will wait until all jobs finish. This is changeable with the `shutdown_timeout` option:

> `shutdown_timeout` (integer) number of seconds to wait for jobs to finish when shutting down before stopping the thread. Defaults to forever: -1. You can also set this with the environment variable GOOD_JOB_SHUTDOWN_TIMEOUT.

As I think you discovered, GoodJob is only single process, so SIGINT/SIGTERM will trigger a shutdown (with Puma, `SIGUSR1` will only restart the individual workers child-process I believe but the parent server process won't restart)
